simpl/incubator/rest.py

```diff
@@ -190,3 +154,41 @@ def wrapped(*args, **kwargs):
                 bottle.abort(400, str(exc))
         return wrapped
     return deco
+
+
+def body(schema=None, types=None, required=False, default=None):
+    """Decorator to parse and validate API body.
+
+    :keyword schema: callable that accepts raw data and returns the coerced (or
+        unchanged) data if it is valid. It should raise an error if the data is
+        not valid.
+    :keyword types: supported content types (default is ['application/json'])
+    :keyword required: if true and no body specified will raise an error.
+    :keyword default: default value to return if no body supplied
+
+    Note: only json types are supported.
+    """
+    if not types:
+        types = ['application/json']
+    if not all('json' in t for t in types):
+        raise NotImplementedError("Only 'json' body supported.")
+
+    def wrap(fxn):
+        """Return a decorated callable."""
+        def wrapped(*args, **kwargs):
+            """Callable to called when the decorated function is called."""
+            body = bottle.request.json
+            if required and not body:
+                bottle.abort(400, "Call body cannot be empty")
+            if body is None:
+                body = default
+            if schema:
+                try:
+                    body = schema(body)
+                except (KeyboardInterrupt, SystemExit):
+                    raise  # don't catch and ignore attempts to end the app
+                except Exception as exc:
+                    bottle.abort(400, str(exc))
+            return fxn(*args, body=body, **kwargs)
```

Comment:
Good question.

Body is singular; our APIs only support one body if it is supplied. And mapping that to one arg is pretty straight-forward, intuitive, and has the benefits discussed above.

I'm not sure the same applies to query params.

To add some use cases to the discussion, here's the list of the most common query params we have in our APIs:
- pagination: `limit` & `offset` (or `marker`)
- `status` - usually a list (supports filtering on more than one) with special keys (ex. `?status=all`)
- text search (`q` param)
- `facet` key for facetted views
- `sort` key - usually multiples also and can be preceded by `-` for descending

Beyond these there are ones we don't use as frequently: `force` for a DELETE or POST, and some extra filters such as `region`, `user` or `tenantId`.

We've put those in kwargs historically, but that has resulted in many functions that look like this:
```python
def route(api_id, region=consts.DEFAULT_REGION,
          limit=consts.DEFAULT_PAGE_SIZE, offset=0,
          status=None):
    do_something()
```
That's pushing the limits of simple/readable IMO, but if you add a couple of route params (ex. id, sub_id) in there it gets too noisy. We could combine some of these into a `query` kwarg as you've done in this PR. We could also use the call context (`simpl.threadlocal`). Or we could continue what we've done above.

I'm not sure. Any suggestions?